ksonnet-gen/ast/ast.go

```diff
@@ -0,0 +1,600 @@
+type LocalBind struct {
+	Variable Identifier
+	Body     Node
+	Fun      *Function
+}
+type LocalBinds []LocalBind
+
+// Local represents local x = e; e.  After desugaring, functionSugar is false.
+type Local struct {
+	NodeBase
+	Binds LocalBinds
+	Body  Node
+}
+
+// ---------------------------------------------------------------------------
+
+// LiteralBoolean represents true and false
+type LiteralBoolean struct {
+	NodeBase
+	Value bool
+}
+
+// ---------------------------------------------------------------------------
+
+// LiteralNull represents the null keyword
+type LiteralNull struct{ NodeBase }
+
+// ---------------------------------------------------------------------------
+
+// LiteralNumber represents a JSON number
+type LiteralNumber struct {
+	NodeBase
+	Value          float64
+	OriginalString string
+}
+
+// ---------------------------------------------------------------------------
+
+type LiteralStringKind int
+
+const (
+	StringSingle LiteralStringKind = iota
+	StringDouble
+	StringBlock
+	VerbatimStringDouble
+	VerbatimStringSingle
+)
+
+func (k LiteralStringKind) FullyEscaped() bool {
+	switch k {
+	case StringSingle, StringDouble:
+		return true
+	case StringBlock, VerbatimStringDouble, VerbatimStringSingle:
+		return false
+	}
+	panic(fmt.Sprintf("Unknown string kind: %v", k))
+}
+
+// LiteralString represents a JSON string
+type LiteralString struct {
+	NodeBase
+	Value       string
+	Kind        LiteralStringKind
+	BlockIndent string
+}
+
+// ---------------------------------------------------------------------------
+
+type ObjectFieldKind int
+
+const (
+	ObjectAssert    ObjectFieldKind = iota // assert expr2 [: expr3]  where expr3 can be nil
+	ObjectFieldID                          // id:[:[:]] expr2
+	ObjectFieldExpr                        // '['expr1']':[:[:]] expr2
+	ObjectFieldStr                         // expr1:[:[:]] expr2
+	ObjectLocal                            // local id = expr2
+)
+
+type ObjectFieldHide int
+
+const (
+	ObjectFieldHidden  ObjectFieldHide = iota // f:: e
+	ObjectFieldInherit                        // f: e
+	ObjectFieldVisible                        // f::: e
+)
+
+// TODO(sbarzowski) consider having separate types for various kinds
+type ObjectField struct {
+	Kind          ObjectFieldKind
+	Hide          ObjectFieldHide // (ignore if kind != astObjectField*)
+	SuperSugar    bool            // +:  (ignore if kind != astObjectField*)
+	MethodSugar   bool            // f(x, y, z): ...  (ignore if kind  == astObjectAssert)
+	Method        *Function
+	Expr1         Node // Not in scope of the object
+	Id            *Identifier
+	Params        *Parameters // If methodSugar == true then holds the params.
+	TrailingComma bool        // If methodSugar == true then remembers the trailing comma
+	Expr2, Expr3  Node        // In scope of the object (can see self).
+
+	// additions to support ksonnet
+	Comment *Comment
+	Inline  bool // print this field inline
```

Comment:
I think this should be called `OneLine` or something. We're not inlining the value of the fields, which would involve evaluating the field itself.

Actually, in an ideal world we'd get to parity generating ksonnet-lib, and remove this member and transition away from having oneline objects at all, as removing this would get us pretty close to having this patch-ready for upstream.